### PR TITLE
Multi-platform support

### PR DESCRIPTION
**Pull request by NoobNotFound**
add multi platform support (wip)

### CODE REVIEW

**Comment by Lamparter** [comment]
This is one of the reasons that UWP is better 😭
I remember multi-platform support was so easy...

File: Emerald.App/Emerald.App/Emerald.App.csproj
Comment (0x5bfa) [inline Emerald.App/Emerald.App/Emerald.App.csproj:11]: Since .NET 8 uses win-(ARCH) uses over win10-(ARCH), you can remove this condition and the right above line.

**Review by 0x5bfa — COMMENTED** [review]
Maybe you guys can use Directory.Build.props to remove duplicated properties among projects?

**Review by 0x5bfa — APPROVED** [review]
Anyway, it looks ok!

> This is one of the reasons that UWP is better 😭
I remember multi-platform support was so easy...

WinAppSdk is more desktop app like. I feel you pain as we’ve used it in FluentHub.

File: Emerald.App/Emerald.App/Emerald.App.csproj
Comment (Lamparter) [reply Emerald.App/Emerald.App/Emerald.App.csproj:11]: I'll implement this suggestion now.

**Comment by NoobNotFound** [comment]
> This is one of the reasons that UWP is better 😭 I remember multi-platform support was so easy...

yeah I agree to @0x5bfa this was UWP at the first (SD Launcher) but it was so difficult to even a run a single Minecraft version because accessing resources is limited. This originally created the option 'Run Minecraft As Admin' because it would solve it at some kind of security risk

**Comment by NoobNotFound** [comment]
btw I planned to rewrite the `Emerald.Core` to `Emerald.CoreX` with a well (i'll try my best) organized code pattern so it would be easy to maintain.
 I was away from the dev for kind of a time so I don't even understand somethings I have written before lol. 

And I'll mostly focused on bringing **Emerald for MacOS** but emerald for Linux will be also created since I hope to use single framework for all. 
`Emerald for Windows` works with no problem at all so I think it'll be a mess to replace `Core` with `CoreX` in `Emerald.WinUI`

BTW what do you guys think should I use for emerald cross platform?
would it be

- MAUI
- MAUI Hybrid (Blazor based MAUI app)
- Blazor with Something like PWA or electron to create a desktop like (idk even this is possible at all lol)
- Uno Platform
- Or something else (Because I think all of these are not just perfect as native WinUI and apple UI (idk whatever they call it)

File: Emerald.App/Emerald.App/Emerald.App.csproj
Comment (NoobNotFound) [reply Emerald.App/Emerald.App/Emerald.App.csproj]: oh, I just noticed that RuntimeIdentifier values has changed after .net 8. wow microsoft is trying to organize more just like me or there must a real reason for it.

**Comment by Lamparter** [comment]
> @RiversideValley/emerald what do you think?

Okay.....
So first things first. In terms of cross-platform, Uno seems like the best way to go as it is the most mature multi platform runtime engine. Blazor/Electron/ASP.NET apps are literally terrible and I wouldn't recommend them in a billion years as they are slow and buggy and the codebase is just messy - desktop apps shouldn't programmed in the same way as websites in my opinion. Plus, ASP.Net Core/Blazor/Electron apps are limited to most OS permission scopes as a website and may be able to open it but wouldn't function as a launcher in the same way that the Minecraft Launcher/Natsurainko.FluentLauncher/Lunar/Badlion/(whatever launchers are being used nowawadays) in terms of being actually able to manage the game on demand - you might as well be developing Emerald as a website that just launches Minecraft via protocols which betrays the true meaning behind what a Minecraft Launcher really is.
Emerald for macOS, on the other hand, will be difficult. I'm not sure whether .NET desktop apps designed and managed by Uno are even supported on macOS.
Then again, this leads me to question, why do we want Emerald to work as a cross-platform app? It's fine to make it universally available across different Windows versions (AMD64, Arm, Arm64, WinRT, etc.) but why do we really want it to function on different platforms? Minecraft on macOS is the only comparable version of the game to that of Windows and I am to suppose that we will have very little userbase who are interested in a Cupertino UI (what Apple's design system is called) Minecraft Launcher that lacks the features or core contributors of other major competitors or even the default Launcher itself? On top of that, Minecraft for Linux is incredibly underdeveloped and resembles Classic Minecraft in its entirety - not a proper game as Minecraft on Windows and macOS appears to be.
Another thing - codebase unifier platforms like Uno and MAUI generally lean towards having a single codebase as that is literally the meaning of the word Uno - it makes no sense to have a separate project for each version as all the apps being able to be built from a singular codebase is the very nature of Uno and MAUI in the first place.
All in all, this is merely speculative but it really does make me question what Emerald really is.

**Comment by NoobNotFound** [comment]
>
> 
> Okay..... So first things first. In terms of cross-platform, Uno seems like the best way to go as it is the most mature multi platform runtime engine. Blazor/Electron/ASP.NET apps are literally terrible and I wouldn't recommend them in a billion years as they are slow and buggy and the codebase is just messy - desktop apps shouldn't programmed in the same way as websites in my opinion.


Yeah I do also prefer Uno over the others, Just listed down the possible ways of doing this. And also, I have never tried either Uno or Blazor with electron(ized) before. So I may need time to get these thing well.


> Plus, ASP.Net Core/Blazor/Electron apps are limited to most OS permission scopes as a website and may be able to open it but wouldn't function as a launcher in the same way that the Minecraft Launcher/Natsurainko.FluentLauncher/Lunar/Badlion/(whatever launchers are being used nowawadays) in terms of being actually able to manage the game on demand - you might as well be developing Emerald as a website that just launches Minecraft via protocols which betrays the true meaning behind what a Minecraft Launcher really is. 

Just to mention, ig some of the popular (legal or cracked) minecraft launchers like TLauncher, Lunarclient uses HTML UI with Java (I'm not sure since I haven't observed any Java based app before, but I think it works like that)

I don't like Electron\Blazor UI based apps because of the performance. The only thing I like about it is we can craft whatever unique UI we want with HTML and CSS (although I'm not a web designer lol)


> Emerald for macOS, on the other hand, will be difficult. I'm not sure whether .NET desktop apps designed and managed by Uno are even supported on macOS. 

There is some built with Uno for macOS, like Devtoys, [Ryujinx](https://github.com/Ryujinx/Ryujinx) (An amazing project I found when searching for emulators, not so sure whether it is poweren by Uno)

>Then again, this leads me to question, why do we want Emerald to work as a cross-platform app? It's fine to make it universally available across different Windows versions (AMD64, Arm, Arm64, WinRT, etc.) but why do we really want it to function on different platforms? 

well Minecraft JAVA Edition is cross-platform, so why not emerald?
Actually I like to bring emerald for macOS since I have a mac tbh.

> Minecraft on macOS is the only comparable version of the game to that of Windows and I am to suppose that we will have very little userbase who are interested in a Cupertino UI (what Apple's design system is called)

Yeah I agree, but I have experienced that annoying user experience when I try to install 3rd party MC launcher on my new mac. 
I had to manually install Java runtime, finding the correct runtime could be challenging not to me but to other normal non-dev peoples like my younger brother who only knows to open emerald and launch minecraft.
Since CMLLib automatically installs required java rumtime which is very easy for normal users.


> Minecraft Launcher that lacks the features or core contributors of other major competitors or even the default Launcher itself? On top of that, Minecraft for Linux is incredibly underdeveloped and resembles Classic Minecraft in its entirety - not a proper game as Minecraft on Windows and macOS appears to be. 

I had only a linux based device before (Which actually came with windows but had to install a linux because it was too old for Windows). I played Minecraft on that device using Tlauncher, I was happy with it. (even though the performence was not so great)
I think Emerald would easily work if it was cross platform thanks to the well-written CMLLib.

> Another thing - codebase unifier platforms like Uno and MAUI generally lean towards having a single codebase as that is literally the meaning of the word Uno - it makes no sense to have a separate project for each version as all the apps being able to be built from a singular codebase is the very nature of Uno and MAUI in the first place. All in all, this is merely speculative but it really does make me question what Emerald really is.

yeah, so do I 😀. Emerald is just a UI for some "custom minecraft launcher" Libraires. made for easy of use, easily download mods using Modrinth API and install and run moded MC version just like TLauncher.


btw what would be the final decision? what about. MAUI? actually I really don't love any of these, I would rather build a native mac app if I knew SwiftUI

**Comment by NoobNotFound** [comment]
wow Ive written a whole essay lol

**Comment by Lamparter** [comment]
Ryujinx isn't powered by Uno afaik

**Comment by Lamparter** [comment]
> well Minecraft JAVA Edition is cross-platform, so why not emerald?
Actually I like to bring emerald for macOS since I have a mac tbh.

Yeah me too

**Comment by Lamparter** [comment]
> btw what would be the final decision? what about. MAUI? actually I really don't love any of these, I would rather build a native mac app if I knew SwiftUI
I know a tiny bit of Swift because it was the second language I learnt after getting into programming but it's very rusty. Also, using an app like Xcode is very unintuitive. Objective-C is also just terrible and I personally think refactoring Emerald via MAUI or Uno is preferable

**Comment by Lamparter** [comment]
Also, I just remembered, MAUI doesn't support Linux so it wouldn't work that way anyway

**Comment by NoobNotFound** [comment]
> Also, I just remembered, MAUI doesn't support Linux so it wouldn't work that way anyway

so, a one solution is gone lol

**Comment by NoobNotFound** [comment]
> Ryujinx isn't powered by Uno afaik


But this feels like WinUI in Uno

https://github.com/user-attachments/assets/b287a5c6-7094-41a8-9bed-543ff97afeda


(don't mind about that time btw)

**Comment by Lamparter** [comment]
Yeah it's a shame since MAUI is very much supported by Microsoft and it would be very nice to get to use it while its new because I imagine that in the near future Microsoft will standardise it as the core universal multi-platform framework for .NET

**Comment by Lamparter** [comment]
> But this feels like WinUI in Uno

Yes it looks like it because they recently replaced the UI with a custom built fluentUI framework for .NET called Avalonia

**Comment by NoobNotFound** [comment]
> Yes it looks like it because they recently replaced the UI with a custom built fluentUI framework for .NET called Avalonia

ahh I know that, how about using it?

**Comment by NoobNotFound** [comment]
> Cross platform WPF 

💀 
never knew that

**Comment by Lamparter** [comment]
> ahh I know that, how about using it?

In Emerald?
That would require rewriting the app from scratch.

**Comment by Lamparter** [comment]
If you think about it Uno has loads of potential.
For example, if you ported an app like FluentHub to Uno you would be able to have a FluentHub mobile, and even a web interface. It's very cool to think about.

**Comment by NoobNotFound** [comment]
> In Emerald? That would require rewriting the app from scratch.

It seems like that. if so, that would be challenging. I can be never done because we don't have that much time. So It should be Emerald.Uno

**Comment by Lamparter** [comment]
What do you mean "we don't have much time"?

**Comment by NoobNotFound** [comment]
> What do you mean "we don't have much time"?

I mean I would try to do that if I had more time to spend on this

**Comment by Lamparter** [comment]
Yes. It seems like Uno is the way to go for now though.

**Comment by Lamparter** [comment]
I give up 😭😭😭
This is practically impossible...

**Comment by NoobNotFound** [comment]
guys I've recreated the store! now it has mods, shaders, plugins, modpacks and resourcepacks support
Screenshot 2024-08-08 at 12 50 11 PM

now it's kind of well organized lol

File: Emerald.App/Emerald.App/Helpers/Updater/Updater.cs
Comment (Lamparter) [inline Emerald.App/Emerald.App/Helpers/Updater/Updater.cs:42]: _Reverside_ 🥴

**Review by Lamparter — CHANGES_REQUESTED** [review]

File: Emerald.App/Emerald.App/Helpers/Updater/Updater.cs
Comment (Lamparter) [reply Emerald.App/Emerald.App/Helpers/Updater/Updater.cs]: Fixed in https://github.com/RiversideValley/Emerald/pull/36/commits/6f9843c002fe7a38d9be1b4e3e5792f6022cc3ae

Comment (NoobNotFound) [reply Emerald.App/Emerald.App/Helpers/Updater/Updater.cs]: > Fixed in [6f9843c](https://github.com/RiversideValley/Emerald/commit/6f9843c002fe7a38d9be1b4e3e5792f6022cc3ae)

thanks

File: Emerald.App/Emerald.App/Helpers/Settings/JSON.cs
Comment (Lamparter) [inline Emerald.App/Emerald.App/Helpers/Settings/JSON.cs:1]: We should try and suppress these warnings 😕 
image

Comment (NoobNotFound) [reply Emerald.App/Emerald.App/Helpers/Settings/JSON.cs:1]: yeah, but we will get them more and more while we are writing the code. so for now, let's just skip it like we have never seen it 🙃 lol

File: Directory.Packages.props
Comment (Lamparter) [inline Directory.Packages.props:14]: @NoobNotFound what is Octokit.NET even used for?

Comment (NoobNotFound) [reply Directory.Packages.props:14]: basically nothing. (actually for the updater which download updates through github releases)

File: Emerald.CoreX/Class1.cs
Comment (Lamparter) [inline Emerald.CoreX/Class1.cs]: The idea of a Core project is that it contains common files across SDKs and implementation projects. I don't understand why we can't just add it's contents to `Emerald.Core` anyway 😕
The name CoreX is in the name of bad form in the first place.. just using Core or calling it OneCore would be better

File: Emerald.CoreX/Emerald.CoreX.csproj
Comment (Lamparter) [inline Emerald.CoreX/Emerald.CoreX.csproj:4]: And even then, it's not particularly any different from the existing Core .net framework project - what's the point?

**Review by Lamparter — CHANGES_REQUESTED** [review]
I mean, yeah the `CoreX` project _does_ establish a difference between winui core library and multi platform library but I don't really see it's point

File: Emerald.App/Emerald.App/Emerald.App.csproj
Comment (Lamparter) [reply Emerald.App/Emerald.App/Emerald.App.csproj]: Done in #42

File: Emerald.CoreX/Class1.cs
Comment (NoobNotFound) [reply Emerald.CoreX/Class1.cs]: The current core is bit outdated and un-organized. I think we should rewrite the Core with new features and a well-organized code. 
It's so hard to replace the current thing at once because every class/thing of `Emerald.Core` depends on another class/thing. That's why I created `Emerald.CoreX`
I also wanted to remove that `Localized.xxxx` thing and `TaskHelper`, because it's kind of thing that should be done by the UI

File: Emerald.CoreX/Emerald.CoreX.csproj
Comment (NoobNotFound) [reply Emerald.CoreX/Emerald.CoreX.csproj:4]: I can agree on that. As I said, after we re-create the CoreX again, we can totally delete the existing `Emerald.Core` and rename `Emerald.CoreX` to  `Emerald.Core`. What do you think?

Comment (Lamparter) [reply Emerald.CoreX/Emerald.CoreX.csproj:4]: LGTM!

File: Emerald.App/Emerald.App/Helpers/Settings/SettingsSystem.cs
Comment (Lamparter) [inline Emerald.App/Emerald.App/Helpers/Settings/SettingsSystem.cs:1]: Emerald.WinUI namespace?

File: Emerald.CoreX/Helpers/FileDownloader.cs
Comment (NoobNotFound) [inline Emerald.CoreX/Helpers/FileDownloader.cs:1]: currently this does not support continuing from interrupted download or pausing, but it should be implemented soon.

File: Emerald.App/Emerald.App/Helpers/Settings/SettingsSystem.cs
Comment (NoobNotFound) [reply Emerald.App/Emerald.App/Helpers/Settings/SettingsSystem.cs:1]: It's not us, it's Microsoft. Let's just skip it and move on lol😅

**Comment by Lamparter** [comment]
I read online that apparently if you do this command:
@sourcery-ai review

**Comment by Lamparter** [comment]
It's meant to do some kind of AI-based PR review?

**Comment by Lamparter** [comment]
@sourcery-ai review

File: Emerald.App/Emerald.App/Helpers/Updater/Updater.cs
Comment (Lamparter) [reply Emerald.App/Emerald.App/Helpers/Updater/Updater.cs:42]: > **🚨 issue (security):** Hard-coded GitHub client ID detected.
> 
> Consider using environment variables or a secure vault to manage sensitive information like client IDs.

This is not a hard-coded client ID.

File: Emerald/Assets/SharedAssets.md
Comment (Lamparter) [reply Emerald/Assets/SharedAssets.md:20]: > **suggestion (documentation):** Remove redundant 'scale-' prefix in directory examples
> 
> The 'scale-' prefix is already included in the file names, so it's not necessary in the directory names. Consider updating these examples for clarity.

@sourcery-ai I don't think I understand. It appears as such in the docs so that it matches that of the source code. Removing the `scale-` prefix would just make it confusing.

Even then, this is an auto-generated file generated by the Uno Platform.

File: Emerald.CoreX/Store/Modrinth/Stores.cs
Comment (NoobNotFound) [reply Emerald.CoreX/Store/Modrinth/Stores.cs:11]: wow, is this free? also thanks for pointing out Ill take a look.

**Comment by Lamparter** [comment]
> Maybe you guys can use Directory.Build.props to remove duplicated properties among projects?

We should do this 👀

**Comment by Lamparter** [comment]
@NoobNotFound let's merge, what do you think? I'll need to review this first but it'll be easier to just commit to main rather than an insanely large PR.

**Review by Lamparter — DISMISSED** [review]

**Review by Lamparter — APPROVED** [review]

File: Emerald/Helpers/Settings/JSON.cs
Comment (NoobNotFound) [reply Emerald/Helpers/Settings/JSON.cs:256]: well the store will be recreated, so these will be obsolute (probably)

File: Emerald.CoreX/Core.cs
Comment (NoobNotFound) [reply Emerald.CoreX/Core.cs:8]: the launcher is still WiP, so will be reviewed in the future

**Pull request merged**
